File: Sources/Chunk/BedrockDBKey.swift
import Foundation

enum BedrockDimension: Int32, CaseIterable {
    case overworld = 0
    case nether = 1
    case end = 2

    var displayName: String {
        switch self {
        case .overworld: return "主世界"
        case .nether: return "下界"
        case .end: return "末地"
        }
    }
}

enum ChunkRecordType: UInt8, CaseIterable {
    case data3D = 0x2b
    case version = 0x2c
    case data2D = 0x2d
    case data2DLegacy = 0x2e
    case subChunk = 0x2f
    case legacyTerrain = 0x30
    case blockEntity = 0x31
    case entity = 0x32
    case pendingTicks = 0x33
    case legacyBlockExtraData = 0x34
    case biomeState = 0x35
    case finalizedState = 0x36
    case conversionData = 0x37
    case borderBlocks = 0x38
    case hardcodedSpawners = 0x39
    case randomTicks = 0x3a
    case checksums = 0x3b
    case generationSeed = 0x3c
    case generatedPreCavesAndCliffsBlending = 0x3d
    case blendingBiomeHeight = 0x3e
    case metadataHash = 0x3f
    case blendingData = 0x40
    case actorDigestVersion = 0x41
    case legacyVersion = 0x76

    var displayName: String {
        switch self {
        case .data3D: return "Data3D"
        case .version: return "Version"
        case .data2D: return "Data2D"
        case .data2DLegacy: return "Data2DLegacy"
        case .subChunk: return "SubChunk"
        case .legacyTerrain: return "LegacyTerrain"
        case .blockEntity: return "BlockEntity"
        case .entity: return "Entity"
        case .pendingTicks: return "PendingTicks"
        case .legacyBlockExtraData: return "LegacyBlockExtraData"
        case .biomeState: return "BiomeState"
        case .finalizedState: return "FinalizedState"
        case .conversionData: return "ConversionData"
        case .borderBlocks: return "BorderBlocks"
        case .hardcodedSpawners: return "HardcodedSpawners"
        case .randomTicks: return "RandomTicks"
        case .checksums: return "Checksums"
        case .generationSeed: return "GenerationSeed"
        case .generatedPreCavesAndCliffsBlending: return "GeneratedPreCavesAndCliffsBlending"
        case .blendingBiomeHeight: return "BlendingBiomeHeight"
        case .metadataHash: return "MetaDataHash"
        case .blendingData: return "BlendingData"
        case .actorDigestVersion: return "ActorDigestVersion"
        case .legacyVersion: return "LegacyVersion"
        }
    }
}

struct ChunkPosition: Hashable {
    let x: Int32
    let z: Int32
    let dimension: Int32
}

struct BedrockDBKey: Hashable {
    let position: ChunkPosition
    let recordType: ChunkRecordType
    let subChunkIndex: Int8?

    static func parse(_ data: Data) -> BedrockDBKey? {
        guard data.count >= 9,
              let x = try? data.littleEndianInt32(at: 0),
              let z = try? data.littleEndianInt32(at: 4) else { return nil }

        if let type = ChunkRecordType(rawValue: data[8]) {
            let subIndex: Int8? = type == .subChunk && data.count >= 10 ? Int8(bitPattern: data[9]) : nil
            return BedrockDBKey(position: ChunkPosition(x: x, z: z, dimension: 0), recordType: type, subChunkIndex: subIndex)
        }
        guard data.count >= 13,
              let dimension = try? data.littleEndianInt32(at: 8),
              let type = ChunkRecordType(rawValue: data[12]) else { return nil }
        let subIndex: Int8? = type == .subChunk && data.count >= 14 ? Int8(bitPattern: data[13]) : nil
        return BedrockDBKey(position: ChunkPosition(x: x, z: z, dimension: dimension), recordType: type, subChunkIndex: subIndex)
    }

    func encoded() -> Data {
        var data = Data()
        data.appendLE(position.x)
        data.appendLE(position.z)
        if position.dimension != 0 { data.appendLE(position.dimension) }
        data.append(recordType.rawValue)
        if recordType == .subChunk, let subChunkIndex = subChunkIndex {
            data.append(UInt8(bitPattern: subChunkIndex))
        }
        return data
    }

    static func subChunk(x: Int32, z: Int32, dimension: Int32, index: Int8) -> Data {
        BedrockDBKey(
            position: ChunkPosition(x: x, z: z, dimension: dimension),
            recordType: .subChunk,
            subChunkIndex: index
        ).encoded()
    }
}


extension BedrockDBKey: CustomStringConvertible {
    var description: String {
        let dimension = BedrockDimension(rawValue: position.dimension)?.displayName ?? "维度 \(position.dimension)"
        if let subChunkIndex = subChunkIndex {
            return "\(dimension) (\(position.x), \(position.z)) \(recordType.displayName) Y=\(subChunkIndex)"
        }
        return "\(dimension) (\(position.x), \(position.z)) \(recordType.displayName)"
    }
}

/// Raw chunk-key matching used by destructive chunk operations.
///
/// Bedrock has both a legacy overworld prefix (`x,z`) and a current
/// dimension-aware prefix (`x,z,dimension`). A one-byte LevelChunkTag follows
/// the prefix; some records append one or two additional bytes (for example a
/// SubChunk Y index). Matching by prefix and bounded suffix length mirrors the
/// Android Blocktopograph `removeFullChunk` implementation and deliberately
/// does not depend on a whitelist of currently known tags.
enum BedrockRawChunkKey {
    static func prefixes(for position: ChunkPosition) -> [Data] {
        var modern = Data()
        modern.appendLE(position.x)
        modern.appendLE(position.z)
        modern.appendLE(position.dimension)

        guard position.dimension == 0 else { return [modern] }
        var legacy = Data()
        legacy.appendLE(position.x)
        legacy.appendLE(position.z)
        return [modern, legacy]
    }

    static func matches(_ key: Data, position: ChunkPosition) -> Bool {
        prefixes(for: position).contains { prefix in
            key.count > prefix.count &&
            key.count <= prefix.count + 3 &&
            key.starts(with: prefix)
        }
    }
}

File: Windows/MCBEEditor.Core/Chunk/BedrockDbKey.cs
using System.Buffers.Binary;

namespace MCBEEditor.Core.Chunk;

public enum BedrockDimension : int
{
    Overworld = 0,
    Nether = 1,
    End = 2
}

public static class BedrockDimensionNames
{
    public static string DisplayName(int dimension) => dimension switch
    {
        0 => "主世界",
        1 => "下界",
        2 => "末地",
        _ => $"维度 {dimension}"
    };
}

public enum ChunkRecordType : byte
{
    Data3D = 0x2b,
    Version = 0x2c,
    Data2D = 0x2d,
    Data2DLegacy = 0x2e,
    SubChunk = 0x2f,
    LegacyTerrain = 0x30,
    BlockEntity = 0x31,
    Entity = 0x32,
    PendingTicks = 0x33,
    LegacyBlockExtraData = 0x34,
    BiomeState = 0x35,
    FinalizedState = 0x36,
    ConversionData = 0x37,
    BorderBlocks = 0x38,
    HardcodedSpawners = 0x39,
    RandomTicks = 0x3a,
    Checksums = 0x3b,
    GenerationSeed = 0x3c,
    GeneratedPreCavesAndCliffsBlending = 0x3d,
    BlendingBiomeHeight = 0x3e,
    MetadataHash = 0x3f,
    BlendingData = 0x40,
    ActorDigestVersion = 0x41,
    LegacyVersion = 0x76
}

public static class ChunkRecordTypeNames
{
    public static string DisplayName(this ChunkRecordType type) => type switch
    {
        ChunkRecordType.Data3D => "Data3D",
        ChunkRecordType.Version => "Version",
        ChunkRecordType.Data2D => "Data2D",
        ChunkRecordType.Data2DLegacy => "Data2DLegacy",
        ChunkRecordType.SubChunk => "SubChunk",
        ChunkRecordType.LegacyTerrain => "LegacyTerrain",
        ChunkRecordType.BlockEntity => "BlockEntity",
        ChunkRecordType.Entity => "Entity",
        ChunkRecordType.PendingTicks => "PendingTicks",
        ChunkRecordType.LegacyBlockExtraData => "LegacyBlockExtraData",
        ChunkRecordType.BiomeState => "BiomeState",
        ChunkRecordType.FinalizedState => "FinalizedState",
        ChunkRecordType.ConversionData => "ConversionData",
        ChunkRecordType.BorderBlocks => "BorderBlocks",
        ChunkRecordType.HardcodedSpawners => "HardcodedSpawners",
        ChunkRecordType.RandomTicks => "RandomTicks",
        ChunkRecordType.Checksums => "Checksums",
        ChunkRecordType.GenerationSeed => "GenerationSeed",
        ChunkRecordType.GeneratedPreCavesAndCliffsBlending => "GeneratedPreCavesAndCliffsBlending",
        ChunkRecordType.BlendingBiomeHeight => "BlendingBiomeHeight",
        ChunkRecordType.MetadataHash => "MetaDataHash",
        ChunkRecordType.BlendingData => "BlendingData",
        ChunkRecordType.ActorDigestVersion => "ActorDigestVersion",
        ChunkRecordType.LegacyVersion => "LegacyVersion",
        _ => $"0x{(byte)type:x2}"
    };
}

public readonly record struct ChunkPosition(int X, int Z, int Dimension)
{
    public string DimensionName => BedrockDimensionNames.DisplayName(Dimension);
    public string CoordinateText => $"({X}, {Z})";
    public override string ToString() => $"{DimensionName} {CoordinateText}";
}

public readonly record struct BedrockDbKey(ChunkPosition Position, ChunkRecordType RecordType, sbyte? SubChunkIndex)
{
    public static bool TryParse(ReadOnlySpan<byte> data, out BedrockDbKey key)
    {
        key = default;
        if (data.Length < 9) return false;
        var x = BinaryPrimitives.ReadInt32LittleEndian(data[..4]);
        var z = BinaryPrimitives.ReadInt32LittleEndian(data.Slice(4, 4));

        if (Enum.IsDefined(typeof(ChunkRecordType), data[8]))
        {
            var type = (ChunkRecordType)data[8];
            sbyte? sub = type == ChunkRecordType.SubChunk && data.Length >= 10 ? unchecked((sbyte)data[9]) : null;
            key = new BedrockDbKey(new ChunkPosition(x, z, 0), type, sub);
            return true;
        }

        if (data.Length < 13 || !Enum.IsDefined(typeof(ChunkRecordType), data[12]))
            return false;
        var dimension = BinaryPrimitives.ReadInt32LittleEndian(data.Slice(8, 4));
        var modernType = (ChunkRecordType)data[12];
        sbyte? modernSub = modernType == ChunkRecordType.SubChunk && data.Length >= 14 ? unchecked((sbyte)data[13]) : null;
        key = new BedrockDbKey(new ChunkPosition(x, z, dimension), modernType, modernSub);
        return true;
    }

    public byte[] Encode()
    {
        var length = 8 + (Position.Dimension == 0 ? 0 : 4) + 1 + (RecordType == ChunkRecordType.SubChunk && SubChunkIndex.HasValue ? 1 : 0);
        var output = new byte[length];
        BinaryPrimitives.WriteInt32LittleEndian(output.AsSpan(0, 4), Position.X);
        BinaryPrimitives.WriteInt32LittleEndian(output.AsSpan(4, 4), Position.Z);
        var offset = 8;
        if (Position.Dimension != 0)
        {
            BinaryPrimitives.WriteInt32LittleEndian(output.AsSpan(offset, 4), Position.Dimension);
            offset += 4;
        }
        output[offset++] = (byte)RecordType;
        if (RecordType == ChunkRecordType.SubChunk && SubChunkIndex.HasValue)
            output[offset] = unchecked((byte)SubChunkIndex.Value);
        return output;
    }

    public static byte[] SubChunk(int x, int z, int dimension, sbyte index)
        => new BedrockDbKey(new ChunkPosition(x, z, dimension), ChunkRecordType.SubChunk, index).Encode();

    public override string ToString()
    {
        var baseText = $"{Position.DimensionName} ({Position.X}, {Position.Z}) {RecordType.DisplayName()}";
        return SubChunkIndex is null ? baseText : $"{baseText} Y={SubChunkIndex.Value}";
    }
}

public static class BedrockRawChunkKey
{
    public static IReadOnlyList<byte[]> Prefixes(ChunkPosition position)
    {
        var modern = new byte[12];
        BinaryPrimitives.WriteInt32LittleEndian(modern.AsSpan(0, 4), position.X);
        BinaryPrimitives.WriteInt32LittleEndian(modern.AsSpan(4, 4), position.Z);
        BinaryPrimitives.WriteInt32LittleEndian(modern.AsSpan(8, 4), position.Dimension);
        if (position.Dimension != 0) return [modern];

        var legacy = new byte[8];
        BinaryPrimitives.WriteInt32LittleEndian(legacy.AsSpan(0, 4), position.X);
        BinaryPrimitives.WriteInt32LittleEndian(legacy.AsSpan(4, 4), position.Z);
        return [modern, legacy];
    }

    public static bool Matches(ReadOnlySpan<byte> key, ChunkPosition position)
    {
        foreach (var prefix in Prefixes(position))
        {
            if (key.Length > prefix.Length && key.Length <= prefix.Length + 3 && key.StartsWith(prefix))
                return true;
        }
        return false;
    }
}
